This is what NSS runs in CI

martinthomson opened the pull request:
If this isn't good - and I'm not qualified to say, nor inclined to review the entire lot - then we have a problem.

@franziskuskiefer or @jallmann might be able to provide input on the changes.

Commit:
Fixed compile error

Commit:
Merged ttauberts changes

Commit:
minor formatting changes

Commit:
Added .travis.yml

Commit:
Added .travis.yml

Commit:
Merge branch 'master' of github.com:jallmann/tls-interop

jallmann committed:
Merge all changes so far.  (#1)

* Implemented cargo tests, Added support for BoringSSL and OpenSSL tests, more changes

Commit:
Minor changes to README and travis_script.sh

franziskuskiefer committed:
Update README.md (#4)

Use travis badge

franziskuskiefer commented:
We've been using other repositories for a while now because no one had access to this one. But it would be great to get access to this one and move back here.
Jonas cleaned up the code a little, enabled travis tests and added openssl interop. Now he's adding more tests. A first batch with ciphersuite tests should get merged today (jallmann/tls-interop/pull/7).

jallmann committed:
Implement cipher tests (#7)

* first steps towards cipher suite tests

* New test cases

* Improved error output to console

* changes to test cases

* Improved console error output

* Basic cipher test functionality ready

* Addressed all revision comments for pull request #7

* Added one more cipher suite to cases.json
fixed BoringSSl and OpenSSL Versions to avoid failures due to version updates.
Changed ecdsa_p521 to ecdsa_p256 in all potential cipher test cases.

Commit:
Minor changes to console output.

ekr commented on travis_script.sh:
```diff
@@ -0,0 +1,32 @@
+cd $ROOT_DIR
+git clone https://github.com/google/boringssl.git
+cd boringssl
+git checkout -q 9af1edbe2201e6c6d58e5e484bf56281d8c751d9
```
Please make this hash a variable.

jallmann replied:
Fixed.

ekr commented on travis_script.sh:
```diff
@@ -0,0 +1,32 @@
+cd $ROOT_DIR
+git clone -q https://github.com/openssl/openssl.git
+cd openssl
+git checkout -q 9e6a32025e6e69949ad3e53a29a0b85f61f30b85
```
Please make this hash a variable as well

jallmann replied:
Fixed.

ekr commented on .travis.yml:
```diff
@@ -0,0 +1,15 @@
+rust:
+  - stable
+env:
+  - LD_LIBRARY_PATH=/home/travis/build/jallmann/dist/Debug/lib:/home/travis/build/jallmann/openssl
```
This seems pretty suspicious.

ekr commented on README.md:
```diff
@@ -1,3 +1,6 @@
+
+[![Build Status](https://travis-ci.org/jallmann/tls-interop.svg?branch=master)](https://travis-ci.org/jallmann/tls-interop)
```
This should not point to jallmann

jallmann replied:
Travis-CI needs to be activated for mozilla/tls-interop first, I think. 
Is there an easy way to make this repository and my fork point ot their respective Travis-CI urls without having to ignore this difference in every future pull request?

jcjones replied:
I am unaware of any. Usually we just end up letting personal branches point back to the Travis-CI of the main project.

I'll work on getting Travis set up for mozilla/tls-interop at https://travis-ci.org/mozilla/tls-interop

ekr commented on all_cipher_cases.json:
```diff
@@ -0,0 +1,942 @@
+{
```
This is not a good way to handle the combinatoric explosion of test cases. Please add a parametrizing operator here that will let you handle things like all the versions and all the cipher suites. You might model this after how our NSS gtests already do this.

jallmann replied:
https://github.com/mozilla/tls-interop/issues/6

jallmann replied:
I will address this in a follow up PR.

ekr commented on run.sh:
```diff
@@ -0,0 +1,69 @@
+#!/usr/bin/env bash
```
I'd like to see this refactored a bit so that you can run interop between any two implementations. I think what you need to do is to have functions which build the client and server arguments and then have some sort of script that combinatorically explodes the options.

jallmann replied:
Opened an issue to address this in the future. 
https://github.com/mozilla/tls-interop/issues/9

ekr commented on src/agent.rs:
```diff
@@ -16,25 +19,43 @@ pub struct Agent {
+    ) -> Result<Agent, i32> {
+        // IPv6 listener by default, IPv4 fallback, unless IPv4 is forced.
+        let addr6 = "[::1]:0".parse().unwrap();
+        let addr4 = "127.0.0.1:0".parse().unwrap();
```
Why not do these lazily when you know the value of ipv4.

jallmann replied:
I'm not sure I understand what you mean. I've personally found this the most readable and compact way to achieve what I described in the comment. The reason to introduce the flag that forces IPv4 was that the OpenSSL-Shim is not IPv6-compatible yet, even in an environment that offers IPv6 and thus lets interop open an IPv6 socket.

ekr replied:
What I'm saying is put the defn of |addr4| and |addr6| inside the appropriate arms of the match.

jcjones replied:
Mmm, since `addr4` is needed in both branches of the match, I'd vote for keeping it as Jonas has it. I think we'd need to change this to a little for loop to have single declarations and accomplish what we're looking for.

ekr commented on src/agent.rs:
```diff
@@ -43,6 +64,16 @@ impl Agent {
+                };
+                match ossl_cipher_format {
+                    true => command.arg(cipher_string_to_ossl(&cipher.to_string())),
+                    false => command.arg(cipher.to_string()),
```
Is there a reason you can't combine these two switches.

jallmann replied:
Honestly, no. I will fix this in a follow up.

ekr commented on src/agent.rs:
```diff
@@ -70,51 +102,54 @@ impl Agent {
-                }
-                _ => return Err(-1),
+
+        match events.iter().next().unwrap().token() {
```
If I am reading this code right, this is a semantic change. The previous code read *every* event, and this just reads the first one. Why is that OK?

jallmann replied:
My reasoning was that there is currently no case in which a second event would ever be processed. The wildcard case in the match statement makes sure the function returns an Error if the first event is neither of type SERVER nor STATUS. Am I wrong?

ekr replied:
It's been a while but when I wrote it, I think I did assume that there might be a second event.

jallmann replied:
I will look into this and opened an issue. 
https://github.com/mozilla/tls-interop/issues/7

ekr commented on src/main.rs:
```diff
@@ -118,16 +124,46 @@ impl Results {
+                };
+                match server {
+                    Some(s) => {
+                        println!("Server: \n{}", String::from_utf8(s.stderr.clone()).unwrap())
```
Can you get these all one line.

jallmann replied:
Fixed.

ekr commented on src/tests.rs:
```diff
@@ -17,3 +24,181 @@ fn flatten_unittest() {
+#[test]
+fn nss_server_vs_ossl_client_all_test_cases() {
+    inner_test_all_cases(ConfigType::OsslClient)
+}
```
Please find some way to parametrize these.

jallmann replied:
Opened an issue to address this in the future. 
https://github.com/mozilla/tls-interop/issues/8

ekr commented on src/tests.rs:
```diff
@@ -1,3 +1,10 @@
+#[cfg(test)]
```
Why do we need the cargo tests that run the interop tests at all.

jallmann replied:
I decided to implement them because it made testing changes during development a lot quicker and more convenient, and because the standard configuration of travis-CI for rust simply runs cargo test.
I realized that there is now basically two ways to achieve the same result, running 'cargo test' or running interop the way it's supposed to be run. Maintaining both options doesn't seem useful, and the run.sh script actually makes the latter almost as convenient as 'cargo test', so I'll focus on improving that instead of the internal rust tests.

ekr reviewed (approved):
I'm going to merge this, but there's a lot of followup work needed in another PR. Can you please prioritize that.

ekr commented:
OK, I was going to merge this but I ran out of time. Jonas, can you please fix the merge conflicts so I can merge?

jallmann committed:
Merge branch 'master' into master

Commit:
Resolved merge conflict

Commit:
Merge branch 'master' of github.com:jallmann/tls-interop

jallmann committed:
Address improvement suggestions (#9)

* Generalized LD_LIB_PATH in .travis.yml

* Minor changes to console output. 

* Resolved merge conflict with mozilla/tls-interop/master

jcjones commented:
It looks like https://github.com/mozilla/tls-interop/pull/5#discussion_r215454044 and https://github.com/mozilla/tls-interop/pull/5#discussion_r218038004 either need follow-ups or issues opened, but both are minor. There are issues #6, #7, #8, and #9 for follow-ups. It looks like this is ready to merge, so I'll do that now.
